pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BalancedInstanceSelector.java

```diff
@@ -51,7 +51,8 @@ public BalancedInstanceSelector(String tableNameWithType, BrokerMetrics brokerMe
 
   @Override
   Map<String, String> select(List<String> segments, int requestId,
-      Map<String, List<String>> segmentToEnabledInstancesMap, Map<String, String> queryOptions) {
+      Map<String, List<String>> segmentToEnabledInstancesMap, Map<String, String> queryOptions,
+      long nowMillis) {
```

Comment:
Why are we passing time here? There should be 2 maps passed to the selector, one for the old segments and one for the new segments

Reply:
Because we want to use the same timestamp to check new segment age for both segment instance assignment and unavailable segment reporting. The other thing is getting clock time is system call. it is not cheap.

The new segments are processed only  in StrictReplicaGroup. I didn't change the behavior for other selector. 

New segments information are directly used through private member in StrictReplicaGroup and thus we don't need passing. 
Please see the implementation for StrictReplicaGroup.

Reply:
Hmm.. I probably should pass new segment info to be thread safe. Let me update the code

Reply:
We want the query time operation to be very light weight. Basically we should process old/new segments and put them into cache before query time

Reply:
PTAL. 
The current logic still checks whether segment is new/old in query time but it doesn't update the cache. Let me know if this is still too complicated. I can remove this to make query time logic lighter.